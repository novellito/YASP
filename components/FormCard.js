import { Card } from 'semantic-ui-react';

const Register = props => {
  return (
    <div className="card-container">
      <Card style={{ width: '400px', padding: '20px' }}>
        <Card.Content>
          <Card.Header content={<h1 className="header">{props.title}</h1>} />
          <Card.Description content={<>{props.type}</>} />
        </Card.Content>
      </Card>
      <style jsx>{`
        .card-container {
          display: flex;
          justify-content: center;
        }
        .header {
          margin: 15px 0;
        }
      `}</style>
    </div>
  );
};

export default Register;
