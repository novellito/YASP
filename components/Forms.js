import { useState } from 'react';
import { connect } from 'react-redux';
import * as actionCreators from '../store/actions/actionCreators';
import Link from 'next/link';
import Router, { withRouter } from 'next/router';
import axios from 'axios';
import { Button, Form, Message } from 'semantic-ui-react';

export const Forms = props => {
  const [emailObj, setEmail] = useState({ email: '', valid: true });
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [invalidCreds, setInvalidCreds] = useState(false);
  const currRoute = props.router.asPath;

  const checkEmailValidity = () => {
    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    if (!emailObj.email) {
      return;
    }
    if (re.test(String(emailObj.email).toLowerCase())) {
      setEmail({ ...emailObj, valid: true });
      return;
    }
    setEmail({ ...emailObj, valid: false });
  };

  const submitForm = async () => {
    if (currRoute === '/register') {
      try {
        const { data } = await axios.post('/api/local/signup', {
          email: emailObj.email,
          username,
          password
        });
        localStorage.setItem('jwt', data.user.token);
        localStorage.setItem('refreshToken', data.user.refreshToken);
        props.setUser({
          id: data.user._id,
          username: data.user.username,
          email: data.user.email
        });
        Router.push('/home');
      } catch (err) {
        console.log(err);
      }
    } else {
      // User is logging in
      const { user } = await props.localLogin(emailObj.email, password);
      if (user) {
        Router.push('/home');
      } else {
        setInvalidCreds(true);
      }
    }
  };

  return (
    <Form error>
      <Form.Field>
        <label htmlFor="email">Email</label>
        <input
          type="text"
          name="email"
          placeholder="test@test.com"
          onBlur={checkEmailValidity}
          onChange={e =>
            setEmail({ ...emailObj, email: e.target.value, valid: true })
          }
          value={emailObj.email}
        />
      </Form.Field>
      {!emailObj.valid ? (
        <Message
          error
          header="Invalid email"
          content="Please enter a valid email address"
        />
      ) : (
        ''
      )}
      {props.register ? (
        <>
          <Form.Field>
            <label htmlFor="username">Username</label>
            <input
              type="text"
              name="username"
              onChange={e => setUsername(e.target.value)}
              value={username}
            />
            {username}
          </Form.Field>
        </>
      ) : (
        ''
      )}
      <Form.Field>
        <label htmlFor="password">Password</label>
        <input
          type="text"
          name="password"
          onChange={e => setPassword(e.target.value)}
          value={password}
        />
        <div className="subtext">
          {invalidCreds ? (
            <>
              <span style={{ color: 'red' }}>Invalid Credentials! </span>
            </>
          ) : (
            ''
          )}
        </div>
        <div className="subtext">
          {currRoute === '/login' ? (
            <>
              <span>Don't have an account? </span>
              <Link href="/register">
                <a>Register here</a>
              </Link>
            </>
          ) : (
            <>
              <span>Already have an account? </span>
              <Link href="/login">
                <a>Log in here</a>
              </Link>
            </>
          )}
        </div>
      </Form.Field>
      <div className="submit-btn">
        <Button color="teal" className="ui button" onClick={submitForm}>
          {currRoute === '/register' ? 'Register' : 'Sign In'}
        </Button>
      </div>
      <style jsx>{`
        .subtext {
          font-size: 0.9em;
          margin: 5px 0;
          text-align: center;
        }
        .submit-btn {
          margin: 10px;
          text-align: center;
        }
      `}</style>
    </Form>
  );
};
const mapDispatchToProps = dispatch => {
  return {
    localLogin: (email, password) =>
      dispatch(actionCreators.localLogin(email, password)),
    setUser: user =>
      dispatch(
        actionCreators.setUser({
          id: user.id,
          email: user.email,
          username: user.username
        })
      ),
    socialLogin: platform => dispatch(actionCreators.socialLogin(platform))
  };
};

export default connect(
  null,
  mapDispatchToProps
)(withRouter(Forms));
