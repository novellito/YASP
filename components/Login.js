import axios from 'axios';

export default class Login extends React.Component {
  state = {
    username: '',
    password: ''
  };

  handleInputChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };
  handleSubmit = async () => {
    try {
      const { data } = await axios.post(
        '/api/local/signup',

        {
          email: this.state.username,
          password: this.state.password
        }
      );
      console.log(data);
    } catch (err) {
      console.log(err);
    }
  };

  render() {
    return (
      <div>
        Username: {/* <form action="">

        </form> */}
        <input
          type="text"
          name="username"
          onChange={e => this.handleInputChange(e)}
          value={this.state.username}
        />
        Password:{' '}
        <input
          type="text"
          name="password"
          onChange={e => this.handleInputChange(e)}
          value={this.state.password}
        />
        <button onClick={this.handleSubmit}>Submit</button>
        <br />
        Username : {this.state.username}
        <br />
        Password : {this.state.password}
      </div>
    );
  }
}
