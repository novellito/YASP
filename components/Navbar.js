import {
  Container,
  Icon,
  Image,
  Menu,
  Sidebar,
  Responsive
} from 'semantic-ui-react';
const rightItems = [
  { as: 'a', content: 'Login', key: 'login' },
  { as: 'a', content: 'Register', key: 'register' }
];
import Link from 'next/link';

const Navbar = props => {
  return (
    <>
      <Menu inverted>
        <Menu.Item>
          <Image size="mini" src="https://react.semantic-ui.com/logo.png" />
        </Menu.Item>

        <Menu.Menu position="right">
          <Menu.Item
            content={
              <Link href="/login">
                <a>Login</a>
              </Link>
            }
          />
          <Menu.Item
            content={
              <Link href="/register">
                <a>Register</a>
              </Link>
            }
          />
        </Menu.Menu>
      </Menu>
    </>
  );
};

export default Navbar;
