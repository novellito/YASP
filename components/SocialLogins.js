import React from 'react';
import { Button } from 'semantic-ui-react';

const SocialLogins = props => {
  return (
    <div className="socials">
      <p>or sign in with</p>
      <Button
        onClick={() => props.login('facebook')}
        color="facebook"
        circular
        icon="facebook"
        disabled={props.disabled}
      />
      <Button
        onClick={() => props.login('twitter')}
        color="twitter"
        circular
        icon="twitter"
        disabled={props.disabled}
      />
      <Button
        onClick={() => props.login('google')}
        style={{ background: '#db4437', color: 'white' }}
        className="google"
        circular
        icon="google"
        disabled={props.disabled}
      />
      <style jsx>{`
        .socials {
          text-align: center;
        }
      `}</style>
    </div>
  );
};

export default SocialLogins;
