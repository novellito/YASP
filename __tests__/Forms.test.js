import { render, cleanup, fireEvent, wait } from 'react-testing-library';
import { Forms } from '../components/Forms';
import MockAdapter from 'axios-mock-adapter';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import axios from 'axios';

import Router from 'next/router';
const mockedRouter = { push: jest.fn(() => 'foo') };
Router.router = mockedRouter;
const mock = new MockAdapter(axios);
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const mockUser = {
  token: '123',
  refreshToken: '134343',
  id: '777',
  email: 'foobar@test.com',
  username: 'foobar'
};

describe('Forms Suite', () => {
  mock.onPost('/api/local/signup').reply(200, {
    user: mockUser
  });
  let store;
  beforeEach(() => {
    store = mockStore({
      login: {
        isLoggedIn: false,
        userId: null,
        email: null,
        username: null
      }
    });
  });

  afterEach(() => {
    localStorage.clear();
    jest.clearAllMocks();
    cleanup();
  });

  // https://github.com/testing-library/react-testing-library/issues/281
  const originalError = console.error;
  beforeAll(() => {
    console.error = (...args) => {
      if (/Warning.*not wrapped in act/.test(args[0])) {
        return;
      }
      originalError.call(console, ...args);
    };
  });

  afterAll(() => {
    console.error = originalError;
  });

  it('should match the login page snapshot', () => {
    const elem = render(<Forms router={{ asPath: '/login' }} />);

    expect(elem).toMatchSnapshot();
  });

  it('should register a new user and navigate to /home', async () => {
    const { container, getByText } = render(
      <Forms router={{ asPath: '/register' }} setUser={jest.fn()} register />
    );
    const spy = jest.spyOn(Storage.prototype, 'setItem');
    const [email, username, password] = container.getElementsByTagName('input');

    fireEvent.change(email, {
      target: { value: 'foo@test.com' }
    });
    fireEvent.change(username, {
      target: { value: 'foo' }
    });
    fireEvent.change(password, {
      target: { value: 'foo123' }
    });

    const registerBtn = getByText('Register');
    fireEvent.click(registerBtn);

    await wait(() => {
      expect(spy).toHaveBeenCalledTimes(2);
      expect(localStorage.length).toEqual(2);
      expect(mockedRouter.push).toHaveBeenCalledWith('/home');
    });
  });

  it('should navigate a user to /home when logging in', async () => {
    const { container, getByText } = render(
      <Forms
        router={{ asPath: '/login' }}
        localLogin={jest.fn(() => ({
          user: {}
        }))}
      />
    );
    const [email, password] = container.getElementsByTagName('input');

    fireEvent.change(email, {
      target: { value: 'foo@test.com' }
    });

    fireEvent.change(password, {
      target: { value: 'foo123' }
    });

    const loginBtn = getByText('Sign In');
    fireEvent.click(loginBtn);

    await wait(() => {
      expect(mockedRouter.push).toHaveBeenCalledWith('/home');
    });
  });

  it('should show invalid credentials when logging in', async () => {
    const { container, getByText } = render(
      <Forms router={{ asPath: '/login' }} localLogin={jest.fn(() => 'no')} />
    );
    const [email, password] = container.getElementsByTagName('input');
    fireEvent.change(email, {
      target: { value: 'foo@test.com' }
    });

    fireEvent.change(password, {
      target: { value: 'foo123' }
    });

    const loginBtn = getByText('Sign In');
    fireEvent.click(loginBtn);

    await wait(() => {
      expect(mockedRouter.push).not.toHaveBeenCalled();
      expect(getByText('Invalid Credentials!')).toBeTruthy();
    });
  });

  it('should indicate that the email entered is INVALID', async () => {
    const { container, getByText } = render(
      <Forms router={{ asPath: '/login' }} />
    );
    const [email] = container.getElementsByTagName('input');

    fireEvent.change(email, {
      target: { value: 'foo' }
    });
    fireEvent.blur(email);

    expect(getByText('Invalid email')).toBeTruthy();
    expect(getByText('Please enter a valid email address')).toBeTruthy();
  });
});
