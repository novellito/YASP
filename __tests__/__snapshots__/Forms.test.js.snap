// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forms Suite should match the login page snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ui error form"
      >
        <div
          class="field"
        >
          <label
            class="jsx-244508578"
            for="email"
          >
            Email
          </label>
          <input
            class="jsx-244508578"
            name="email"
            placeholder="test@test.com"
            type="text"
            value=""
          />
        </div>
        
        
        <div
          class="field"
        >
          <label
            class="jsx-244508578"
            for="password"
          >
            Password
          </label>
          <input
            class="jsx-244508578"
            name="password"
            type="text"
            value=""
          />
          <div
            class="jsx-244508578 subtext"
          />
          <div
            class="jsx-244508578 subtext"
          >
            <span
              class="jsx-244508578"
            >
              Don't have an account? 
            </span>
            <a
              class="jsx-244508578"
              href="/register"
            >
              Register here
            </a>
          </div>
        </div>
        <div
          class="jsx-244508578 submit-btn"
        >
          <button
            class="ui teal button ui button"
          >
            Sign In
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ui error form"
    >
      <div
        class="field"
      >
        <label
          class="jsx-244508578"
          for="email"
        >
          Email
        </label>
        <input
          class="jsx-244508578"
          name="email"
          placeholder="test@test.com"
          type="text"
          value=""
        />
      </div>
      
      
      <div
        class="field"
      >
        <label
          class="jsx-244508578"
          for="password"
        >
          Password
        </label>
        <input
          class="jsx-244508578"
          name="password"
          type="text"
          value=""
        />
        <div
          class="jsx-244508578 subtext"
        />
        <div
          class="jsx-244508578 subtext"
        >
          <span
            class="jsx-244508578"
          >
            Don't have an account? 
          </span>
          <a
            class="jsx-244508578"
            href="/register"
          >
            Register here
          </a>
        </div>
      </div>
      <div
        class="jsx-244508578 submit-btn"
      >
        <button
          class="ui teal button ui button"
        >
          Sign In
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
